Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,289 @@
+        CNIOOpenSSL_BIO_set_init(self.bioPtr, 1)
+        CNIOOpenSSL_BIO_set_shutdown(self.bioPtr, 1)
+    }
+
+    deinit {
+        // On deinit we need to drop our reference to the BIO, and also ensure that it doesn't hold any
+        // pointers to this object anymore. We also want to dirty our own ref to the BIO.
+        CNIOOpenSSL_BIO_set_data(self.bioPtr, nil)
+        BIO_free(self.bioPtr)
+        self.bioPtr = UnsafeMutablePointer(bitPattern: 0xdeadbeef)!
+    }
+
+    /// Called to obtain the outbound ciphertext written by OpenSSL.
+    ///
+    /// This function obtains a buffer of ciphertext that needs to be written to the network. In a
+    /// normal application, this should be obtained on a call to `flush`. If no bytes have been flushed
+    /// to the network, then this call will return `nil` rather than an empty byte buffer, to help signal
+    /// that the `write` call should be elided.
+    ///
+    /// - returns: A buffer of ciphertext to send to the network, or `nil` if no buffer is available.
+    func outboundCiphertext() -> ByteBuffer? {
+        guard self.outboundBuffer.readableBytes > 0 else {
+            // No data to send.
+            return nil
+        }
+
+        /// Once we return from this function, we want to account for the bytes we've handed off.
+        defer {
+            self.outboundBuffer.moveReaderIndex(to: self.outboundBuffer.writerIndex)
+        }
+
+        return self.outboundBuffer
+    }
+
+    /// Stores a buffer received from the network for delivery to OpenSSL.
+    ///
+    /// Whenever a buffer is received from the network, it is passed to the BIO via this function
+    /// call. Only one buffer may be passed to OpenSSL at any one time: once a buffer is passed, it
+    /// is expected to be cleared immediately. If it is not cleared, this is an application error and
+    /// must be resolved.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer of ciphertext bytes received from the network.
+    func receiveFromNetwork(buffer: ByteBuffer) {
+        precondition(self.inboundBuffer == nil, "Did not flush inbound bytes to OpenSSL")
+        self.inboundBuffer = buffer
+    }
+
+    /// OpenSSL has requested to read ciphertext bytes from the network.
+    ///
+    /// This function is invoked whenever OpenSSL is looking to read data.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer for NIO to copy bytes into.
+    /// - returns: The number of bytes we have copied.
+    fileprivate func sslRead(buffer: UnsafeMutableBufferPointer<CChar>) -> CInt {
+        guard var inboundBuffer = self.inboundBuffer else {
+            // We have no bytes to read. Mark this as "needs read retry".
+            CNIOOpenSSL_BIO_set_retry_read(self.bioPtr)
+            return -1
+        }
+
+        let bytesToCopy = min(buffer.count, inboundBuffer.readableBytes)
+        _ = inboundBuffer.readWithUnsafeReadableBytes { bytePointer in
```

Comment:
does this function not allow to discard the return value? It should...

Reply:
It does not.